File: JUC/src/main/java/join/JoinDemo.java
package join;

/**
 * this is a study project
 *
 * @author iszhaoy
 * @version 1.0
 * @date 2020/6/4 16:34
 */
public class JoinDemo {

    public static void main(String[] args) {
        Thread scanDouyinThread = new Thread(new ScanDouYin());
        Thread sleep = new Thread(new ReadySleep(scanDouyinThread));
        sleep.start();
        scanDouyinThread.start();
    }
}

class ScanDouYin implements Runnable {
    // ms
    private int time = 10000;

    @Override
    public void run() {
        System.out.println("开始刷抖音喽~~~");
        try {
            Thread.sleep(time);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        System.out.println("刷美了~~~");
    }
}

class ReadySleep implements Runnable {

    private Thread scanDouYin;

    public ReadySleep(Thread scanDouYin) {
        this.scanDouYin = scanDouYin;
    }

    @Override
    public void run() {
        try {
            System.out.println("准备睡觉了。。。。");
            scanDouYin.join();
            System.out.println("睡觉吧。。。。");
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: spark-demo/src/main/java/com/iszhaoy/join/JoinDemo.scala
package com.iszhaoy.join

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

/**
 * this is a study project
 *
 * @author iszhaoy
 * @date 2020/6/28 10:07
 * @version 1.0
 */
object JoinDemo {
  def main(args: Array[String]): Unit = {

    val conf: SparkConf = new SparkConf()
      .setMaster("local[*]")
      .setAppName("acc")

    val sc: SparkContext = SparkContext.getOrCreate(conf)
    val leftRDD: RDD[(Int, String)] = sc.makeRDD(List((1, "one"), (2, "two"), (3, "three")), 2)
    val rightRDD: RDD[(Int, String)] = sc.makeRDD(List((1, "一"), (2, "二"), (4, "四")), 2)

    val joinRDD: RDD[(Int, (String, String))] = leftRDD.join(rightRDD)
    joinRDD.collect()
      .foreach(println)
    val leftJoinRDD: RDD[(Int, (String, Option[String]))] = leftRDD.leftOuterJoin(rightRDD)

    leftJoinRDD.map {
      case (key, (en, zh)) => (key, en, zh.orNull)
    }.collect().foreach(println)


    val rightJoinRDD: RDD[(Int, (Option[String], String))] = leftRDD.rightOuterJoin(rightRDD)

    rightJoinRDD.map {
      case (key, (en, zh)) => (key, en.orNull, zh)
    }.collect().foreach(println)

    leftRDD.fullOuterJoin(rightRDD).map {
      case (key, (en, zh)) => (key, en.orNull, zh.orNull)
    }
      .collect().foreach(println)
    sc.stop()

  }
}